stl/inc/ranges

```diff
@@ -6837,6 +6837,602 @@ namespace ranges {
+            (is_nothrow_move_constructible_v<_ViewTypes> && ...)) // strengthened
+            : _M_Views(_STD move(_Views)...) {}
+
+        _NODISCARD constexpr auto begin() noexcept(
+            noexcept(_Zip::_Tuple_transform(_RANGES begin, _M_Views))) // strengthened
+            requires(!(_Simple_view<_ViewTypes> && ...)) {
+            return _Iterator<false>{_Zip::_Tuple_transform(_RANGES begin, _M_Views)};
+        }
+
+        _NODISCARD constexpr auto begin() const
+            noexcept(noexcept(_Zip::_Tuple_transform(_RANGES begin, _M_Views))) // strengthened
+            requires(range<const _ViewTypes>&&...) {
+            return _Iterator<true>{_Zip::_Tuple_transform(_RANGES begin, _M_Views)};
+        }
+
+        _NODISCARD constexpr auto end() noexcept(
+            noexcept(_Sentinel<false>{_Zip::_Tuple_transform(_RANGES end, _M_Views)}) && noexcept(
+                _Iterator<false>{_Zip::_Tuple_transform(_RANGES end, _M_Views)}) && noexcept(_RANGES next(begin(),
+                iter_difference_t<_Iterator<false>>{_Get_signed_size()}))) // strengthened
+            requires(!(_Simple_view<_ViewTypes> && ...)) {
+            if constexpr (!_Zip::_Is_common<_ViewTypes...>) {
+                return _Sentinel<false>{_Zip::_Tuple_transform(_RANGES end, _M_Views)};
+            } else if constexpr ((random_access_range<_ViewTypes> && ...)) {
+                return _RANGES next(begin(), iter_difference_t<_Iterator<false>>{_Get_signed_size()});
+            } else {
+                return _Iterator<false>{_Zip::_Tuple_transform(_RANGES end, _M_Views)};
+            }
+        }
+
+        _NODISCARD constexpr auto end() const
+            noexcept(noexcept(_Sentinel<true>{_Zip::_Tuple_transform(_RANGES end, _M_Views)}) && noexcept(
+                _Iterator<true>{_Zip::_Tuple_transform(_RANGES end, _M_Views)}) && noexcept(_RANGES next(begin(),
+                iter_difference_t<_Iterator<true>>{_Get_signed_size()}))) // strengthened
+            requires(range<const _ViewTypes>&&...) {
+            if constexpr (!_Zip::_Is_common<const _ViewTypes...>) {
+                return _Sentinel<true>{_Zip::_Tuple_transform(_RANGES end, _M_Views)};
+            } else if constexpr ((random_access_range<const _ViewTypes> && ...)) {
+                return _RANGES next(begin(), iter_difference_t<_Iterator<true>>{_Get_signed_size()});
+            } else {
+                return _Iterator<true>{_Zip::_Tuple_transform(_RANGES end, _M_Views)};
+            }
+        }
+
+        _NODISCARD constexpr auto size() noexcept(
+            noexcept(_Zip::_Tuple_for_each(_RANGES size, _M_Views))) // strengthened
+            requires(sized_range<_ViewTypes>&&...) {
+            using _Common_size_type          = common_type_t<range_size_t<_ViewTypes>...>;
+            using _Common_unsigned_size_type = _Make_unsigned_like_t<_Common_size_type>;
+
+            auto _Smallest_view_size_value = numeric_limits<_Common_unsigned_size_type>::max();

```

Comment:
Although it is technically forbidden by the Standard, `min` and `max` are commonly defined as function-like macros (notably by `<Windows.h>` for ancient legacy reasons). Thus in the STL we must defend all usage of `min` and `max` in product code. Fortunately, extra parentheses are sufficient to prevent the preprocessor from expanding function-like macros, so this should be `(numeric_limits<_Common_unsigned_size_type>::max)()`. Occurs below.

Reply:
Alternatively, careful use of the `NOMINMAX` macro can be used to avoid this behavior, but I like the simplicity and safety of your approach more. I just wanted to point that out since the existence of that macro doesn't seem to get enough attention, but to be perfectly clear, I'd be wary about using it in a project like this. Otherwise, you could easily break other code which relies on those macros if you don't restore the original definition (e.g., with `#pragma push_macro` and `#pragma pop_macro`)!

Reply:
Note that while `NOMINMAX` can be useful for applications, it is not applicable to the STL, because users can include `<Windows.h>` before any STL header (so even if the STL tried to define `NOMINMAX`, it would be too late). That is, `NOMINMAX` only affects whether `<Windows.h>` defines the macros - it does not affect their expansion if they've already been defined.

We use `push_macro`/`pop_macro` to defend ourselves against (evil) macroized `new` as when that's macroized (e.g. by MFC :crying_cat_face:) it's an object-like macro. We could have done that for `min` and `max` too, but historically we chose the more "portable" parenthesis defense.